### PR TITLE
Choose status which abort the flow build.

### PR DESCRIPTION
**Pull request by ueberhammDesign**
You can choose the status (Success, Unstable or Failure), which abort the flow build. It also contains, that the build can not be 'better' than any build inside the flow build.

### CODE REVIEW

File: src/main/groovy/com/cloudbees/plugins/flow/FlowDSL.groovy [hunk 1]
```diff
@@ -201,7 +202,8 @@ public class FlowDelegate {
      * Check flow status and stop if unexpected failure is detected
      */
     private void statusCheck() {
-        if (flowRun.state.result.isWorseThan(SUCCESS)) {
+        if (flowRun.state.result.isWorseThan(Result.fromString(flowRun.getAbortWhenWorseThan()))) {
+            println("Abort execution, because on of the last builds is worse than " + flowRun.getAbortWhenWorseThan())
```
Comment (dnozay) [inline src/main/groovy/com/cloudbees/plugins/flow/FlowDSL.groovy]: typo

File: src/main/groovy/com/cloudbees/plugins/flow/FlowDSL.groovy [hunk 2]
```diff
@@ -454,7 +456,15 @@ public class FlowDelegate {
             pool.awaitTermination(1, TimeUnit.DAYS)
             current_state.lastCompleted =lastCompleted
         } finally {
-            flowRun.state = current_state
+            def current_result = current_state.result
+            if(current_result.isWorseOrEqualTo(flowRun.result)) {
```
Comment (dnozay) [inline src/main/groovy/com/cloudbees/plugins/flow/FlowDSL.groovy]: syntax not consistent.

File: src/main/groovy/com/cloudbees/plugins/flow/FlowDSL.groovy [hunk 3]
```diff
@@ -454,7 +456,15 @@ public class FlowDelegate {
...
+                flowRun.state = current_state
+            } else {
+                def flowRunResult = flowRun.result
+                flowRun.state = current_state
+                flowRun.result = flowRunResult
```
Comment (dnozay) [inline src/main/groovy/com/cloudbees/plugins/flow/FlowDSL.groovy]: isn't this a no-op?

File: src/main/java/com/cloudbees/plugins/flow/BuildFlow.java [hunk 4]
```diff
@@ -97,11 +89,20 @@ public void setDslFile(String dslFile) {
...
+    public void setAbortWhenWorseThan(String abortWhenWorseThan) {
+        this.abortWhenWorseThan = abortWhenWorseThan;
+    }
+
     @Override
     protected void submit(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException, FormException {
         super.submit(req, rsp);
         JSONObject json = req.getSubmittedForm();
         this.buildNeedsWorkspace = json.containsKey("buildNeedsWorkspace");
+        this.abortWhenWorseThan = json.getString("abortWhenWorseThan");
```
Comment (dnozay) [inline src/main/java/com/cloudbees/plugins/flow/BuildFlow.java]: default for old job configs? Is it backwards compatible?

File: src/main/java/com/cloudbees/plugins/flow/FlowRun.java [hunk 5]
```diff
@@ -103,13 +101,17 @@ private void setup(BuildFlow job) {
 
     /* package */ Run waitForCompletion(JobInvocation job) throws ExecutionException, InterruptedException {
         job.waitForCompletion();
-        getState().setResult(job.getResult());
+        if(getState().getResult().isBetterOrEqualTo(job.getResult())) {
```
Comment (dnozay) [inline src/main/java/com/cloudbees/plugins/flow/FlowRun.java]: 1. syntax not consistent.
2. how do we turn off this behavior?

File: src/main/resources/com/cloudbees/plugins/flow/BuildFlow/help-abortWhenWorserThan.jelly [hunk 6]
```diff
@@ -0,0 +1,35 @@
+<!--
+  ~ The MIT License
+  ~
+  ~ Copyright (c) 2013, CloudBees, Inc., Nicolas De Loof.
```
Comment (dnozay) [inline src/main/resources/com/cloudbees/plugins/flow/BuildFlow/help-abortWhenWorserThan.jelly]: improper attribution.

File: src/test/groovy/com/cloudbees/plugins/flow/BuildTest.groovy [hunk 7]
```diff
@@ -148,6 +148,132 @@ class BuildTest extends DSLTestCase {
         println flow.jobsGraph.edgeSet()
     }
 
+    public void testSequentialBuildsWithUnstableByDefault() {
+        def jobs = createJobs(["job1", "job2", "job3"])
+        def willBeUnstable = createUnstableJob("willBeUnstable")
+        def notRan = createJob("notRan")
```
Comment (dnozay) [inline src/test/groovy/com/cloudbees/plugins/flow/BuildTest.groovy]: willNotRun sounds more correct (mixing past / future tense).

File: src/test/groovy/com/cloudbees/plugins/flow/BuildTest.groovy [hunk 8]
```diff
@@ -148,6 +148,132 @@ class BuildTest extends DSLTestCase {
...
+        assertDidNotRun(notRan)
+        assert UNSTABLE == flow.result
+        println flow.jobsGraph.edgeSet()
+    }
+
+    public void testSequentialBuildsWithUnstableBySucess() {
+        def jobs = createJobs(["job1", "job2", "job3"])
+        def willBeUnstable = createUnstableJob("willBeUnstable")
+        def notRan = createJob("notRan")
+        def flow = runWithAbortBySuccess("""
```
Comment (dnozay) [inline src/test/groovy/com/cloudbees/plugins/flow/BuildTest.groovy]: need refactor.

File: src/test/groovy/com/cloudbees/plugins/flow/BuildTest.groovy [hunk 9]
```diff
@@ -148,6 +148,132 @@ class BuildTest extends DSLTestCase {
...
+        assertDidNotRun(notRan)
+        assert UNSTABLE == flow.result
+        println flow.jobsGraph.edgeSet()
+    }
+
+    public void testSequentialBuildsWithFailureBySucess() {
+        def jobs = createJobs(["job1", "job2", "job3"])
+        def willFail = createFailJob("willFail")
+        def notRan = createJob("notRan")
+        def flow = runWithAbortBySuccess("""
```
Comment (dnozay) [inline src/test/groovy/com/cloudbees/plugins/flow/BuildTest.groovy]: need refactor.

**Comment by dnozay** [comment]
awesome that there are tests! :+1: please rebase and make sure that the feature is opt-in only. thanks a lot.

File: src/main/resources/com/cloudbees/plugins/flow/BuildFlow/help-useAbortWhenWorseThan.jelly [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt" xmlns:p="/lib/hudson/project">
+    <div>
+       <div>
+           Abort the build, when one single step in the build process is worster than your value.
```
Comment (dnozay) [inline src/main/resources/com/cloudbees/plugins/flow/BuildFlow/help-useAbortWhenWorseThan.jelly]: worse

File: src/main/java/com/cloudbees/plugins/flow/FlowRun.java [hunk 11]
```diff
@@ -103,13 +103,21 @@ private void setup(BuildFlow job) {
 
     /* package */ Run waitForCompletion(JobInvocation job) throws ExecutionException, InterruptedException {
         job.waitForCompletion();
-        getState().setResult(job.getResult());
+        if(useAbortWhenWorseThan && getState().getResult().isBetterOrEqualTo(job.getResult())) {
```
Comment (dnozay) [inline src/main/java/com/cloudbees/plugins/flow/FlowRun.java]: syntax for if conditions should be consistent; please use `if (condition) {`

**Comment by dnozay** [comment]
once you are done with your changes, I would also appreciate if you could squash relevant changes together. thank you!

**Comment by ueberhammDesign** [comment]
All changes squashed together.

File: src/main/groovy/com/cloudbees/plugins/flow/FlowDSL.groovy [hunk 12]
```diff
@@ -26,7 +26,7 @@
 package com.cloudbees.plugins.flow
 
 import hudson.AbortException
-import hudson.console.HyperlinkNote
+import hudson.console.ModelHyperlinkNote
```
Comment (dnozay) [inline src/main/groovy/com/cloudbees/plugins/flow/FlowDSL.groovy]: this doesn't look right. did you squash commits that were already merged?

**Pull request closed without merging**